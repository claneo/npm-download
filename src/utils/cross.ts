export default function cross<T1, T2>(arr1: T1[], arr2: T2[]): [T1, T2][] {
  const result: [T1, T2][] = [];
  for (const v1 of arr1) {
    for (const v2 of arr2) {
      result.push([v1, v2]);
    }
  }
  return result;
}
