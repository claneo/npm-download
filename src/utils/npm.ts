import cacache from 'cacache';
import chalk from 'chalk';
import fs from 'fs';
import libNpmConfig from 'libnpmconfig';
import { publish as libNpmPublish } from 'libnpmpublish';
import npa from 'npm-package-arg';
import pickManifest from 'npm-pick-manifest';
import regFetch from 'npm-registry-fetch';
import osenv from 'osenv';
import pacote from 'pacote';
import path from 'path';
import ssri from 'ssri';
import downloadedFilename from './downloadedFilename';

process.env['FORCE_COLOR'] = chalk.level.toString();

export const config = libNpmConfig
  .read({
    tmp: osenv.tmpdir(),
    cache: path.join(require('pacote/lib/util/cache-dir')(), '_cacache'),
  })
  .toJSON();

export const setRegistry = (registry: 'npm' | 'taobao' | string) => {
  if (registry === 'npm') registry = 'https://registry.npmjs.org/';
  if (registry === 'taobao') registry = 'https://registry.npm.taobao.org/';
  config.registry = registry;
};

export const distTag = {
  ls: (pkg: string) =>
    regFetch.json(
      `/-/package/${encodeURIComponent(pkg)}/dist-tags`,
      config,
    ) as Promise<Record<string, string>>,
  add: (pkg: string, tag: string, version: string) =>
    regFetch.json(
      `/-/package/${encodeURIComponent(pkg)}/dist-tags/${encodeURIComponent(
        tag,
      )}`,
      {
        ...config,
        method: 'PUT',
        body: JSON.stringify(version),
        headers: {
          'content-type': 'application/json',
        },
      },
    ) as Promise<Record<string, string>>,
  rm: (pkg: string, tag: string) =>
    regFetch.json(
      `/-/package/${encodeURIComponent(pkg)}/dist-tags/${encodeURIComponent(
        tag,
      )}`,
      {
        ...config,
        method: 'DELETE',
      },
    ) as Promise<Record<string, string>>,
};

export const publish = (pkg: string, tag = 'latest') =>
  new Promise(resolve =>
    cacache.tmp.withTmp(config.tmp, { tmpPrefix: 'fromPackage' }, tmp => {
      const cb = (async () => {
        const extracted = path.join(tmp, 'package');
        const target = path.join(extracted, 'package.json');
        await pacote.extract(pkg, extracted);
        const packageJsonString = await fs.promises.readFile(target, 'utf8');
        const packageJson = JSON.parse(packageJsonString);
        delete packageJson.publishConfig;
        return libNpmPublish(packageJson, fs.createReadStream(pkg), {
          ...config,
          tag,
        });
      })();
      resolve(cb);
      return cb;
    }),
  );

export const view = async (pkg: string) => {
  const packument = await pacote.packument(pkg, {
    cache: config.cache,
    registry: config.registry,
  });
  const manifest = pickManifest(packument, npa(pkg).fetchSpec);
  if (!manifest.dist.integrity && manifest.dist.shasum)
    manifest.dist.integrity = ssri
      .fromHex(manifest.dist.shasum, 'sha1')
      .toString();

  return { ...manifest, 'dist-tags': packument['dist-tags'] };
};

export const pack = async (pkg: string, dir: string = process.cwd()) => {
  const { registry, cache } = config;
  const packageManifest = await view(pkg);
  return (
    pacote.tarball
      // @ts-ignore
      .file(pkg, path.join(dir, downloadedFilename(pkg)), {
        ...config,
        registry,
        cache,
        resolved: packageManifest.dist.tarball,
        integrity: packageManifest.dist.integrity,
      })
      .catch((e: any) => {
        console.log(packageManifest, e);
        process.exit();
      })
  );
};
