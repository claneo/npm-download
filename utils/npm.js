const path = require('path');
const pirates = require('pirates');
const spawn = require('cross-spawn');
const Promise = require('bluebird');
const {
  Worker,
  isMainThread,
  workerData,
  parentPort,
} = require('worker_threads');
const chalk = require('chalk');

process.env['FORCE_COLOR'] = chalk.level.toString();

let npmPath;

try {
  npmPath = require.resolve('npm');
} catch (error) {
  if (process.platform === 'win32') {
    dirname = path.join(process.env.APPDATA, 'npm');
    try {
      npmPath = require.resolve(
        path.join(process.env.APPDATA, 'npm/node_modules/npm'),
      );
    } catch (error) {
      npmPath = require.resolve(
        path.join(path.dirname(process.execPath), 'node_modules/npm'),
      );
    }
  } else {
    npmPath = require.resolve(
      path.join(path.dirname(process.execPath), '../lib/node_modules/npm'),
    );
  }
}

const npm = require(npmPath);

const load = extraConfig =>
  new Promise(resolve =>
    npm.load(
      {
        silent: true,
        json: true,
        loglevel: 'silent',
        audit: false,
        ...extraConfig,
      },
      resolve,
    ),
  );

// module.exports.load = (config = {}) =>
//   new Promise(resolve => npm.load(config, resolve));
module.exports.view = package =>
  load().then(
    () =>
      new Promise((resolve, reject) =>
        npm.commands.view([package], true, (_, result) => {
          const values = Object.values(result);
          if (!values) reject();
          else resolve(values[0]);
        }),
      ),
  );
module.exports.distTag = {
  ls: pkg =>
    load().then(
      () =>
        new Promise(resolve =>
          npm.commands.distTag(['ls', pkg], (err, data) => {
            if (err) reject(err);
            else resolve(data);
          }),
        ),
    ),
  add: (pkg, tag, version) =>
    load().then(
      () =>
        new Promise((resolve, reject) =>
          npm.commands.distTag(
            ['add', `${pkg}@${version}`, tag],
            (err, data) => {
              if (err) reject(err);
              else resolve(data);
            },
          ),
        ),
    ),
  rm: (pkg, tag) =>
    load().then(
      () =>
        new Promise((resolve, reject) =>
          npm.commands.distTag(['rm', pkg, tag], (err, data) => {
            if (
              !err ||
              (typeof err.message === 'string' &&
                err.message.includes('is not a dist-tag on'))
            )
              resolve(data);
            else reject(err);
          }),
        ),
    ),
};
if (isMainThread) {
  module.exports.publish = (packages, isLatest) =>
    new Promise((resolve, reject) => {
      const worker = new Worker(__filename, {
        workerData: { packages, isLatest, action: 'publish' },
      });
      worker.addListener('message', resolve);
    });
} else {
  const { packages, isLatest, action } = workerData;
  const publishOnce = pkg =>
    new Promise((resolve, reject) =>
      npm.commands.publish(['./download/' + pkg], v => {
        if (v instanceof Error) reject(v);
        else resolve(v);
      }),
    );
  const publish = pkg => {
    if (isLatest) console.log(`uploading latest package: ${pkg}`);
    else console.log(`uploading non-latest package: ${pkg}`);
    return publishOnce(pkg).catch(() => {
      console.log(chalk.red(`upload ${pkg} failed, retry`));
      return publish(pkg);
    });
  };
  if (action === 'publish')
    load({ tag: isLatest ? 'latest' : 'false' })
      .then(() =>
        Promise.map(packages, pkg => publish(pkg, isLatest), {
          concurrency: 10,
        }),
      )
      .then(() => {
        parentPort.postMessage('ok');
        // process.exit();
      });
}
module.exports.pack = (pkg, dir) =>
  load().then(
    () =>
      new Promise((resolve, reject) =>
        npm.commands.pack([pkg], true, (err, data) => {
          if (err) reject(err);
          else resolve(data);
        }),
      ),
  );
// module.exports.install = packages =>
//   load.then(() => {
//     if (!Array.isArray(packages)) packages = [packages];
//     return new Promise((resolve, reject) => {
//       const log = console.log;
//       console.log = () => {};
//       new npm.commands.install.Installer(
//         path.join(__dirname, '../temp'),
//         true,
//         packages
//       ).run((err, installed) => {
//         console.log = log;
//         if (err) reject(err);
//         resolve(installed.map(item => item[0]));
//       });
//     });
//   });
module.exports.install = packages =>
  new Promise((resolve, reject) => {
    if (!Array.isArray(packages)) packages = [packages];
    const install = spawn('npm', [
      '--json',
      '-s',
      'install',
      '--prefix',
      path.join(__dirname, '../temp'),
      '--dry-run',
      ...packages,
    ]);
    let result = '';
    install.stdout.on('data', data => {
      result += data.toString();
    });
    install.stdout.on('close', () => {
      resolve(JSON.parse(result));
    });
    install.stderr.on('data', data => {
      console.log(data.toString());
      reject();
    });
  });

pirates.addHook(
  code => {
    return code.replace(
      `BB.promisify(require('read-package-json'))`,
      `BB.promisify((...params) => {
        const cb = params.pop();
        require('read-package-json')(...params, (err, json) => {
          delete json.publishConfig;
          cb(err, json);
        });
      })`,
    );
  },
  {
    ignoreNodeModules: false,
    matcher: filename =>
      filename === path.join(path.dirname(npmPath), 'publish.js'),
  },
);

pirates.addHook(
  code => {
    return code.replace(`npmConfig()`, `require('./config/figgy-config.js')()`);
  },
  {
    ignoreNodeModules: false,
    matcher: filename =>
      filename === path.join(path.dirname(npmPath), 'pack.js'),
  },
);
