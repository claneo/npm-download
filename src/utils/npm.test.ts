import { omit } from 'lodash';
import osenv from 'osenv';
import path from 'path';
import cross from './cross';
import { config, distTag, setRegistry, view, publish } from './npm';

describe('npm config', () => {
  it('cache', () => {
    expect(config.cache).toBe(
      path.join(require('pacote/lib/util/cache-dir')(), '_cacache'),
    );
  });
  it('tmp', () => {
    expect(config.tmp).toBe(osenv.tmpdir());
  });
});

describe('npm publishSingle', () => {
  it('packageJson', async () => {
    setRegistry('http://localhost:8081/repository/npm/');
    const data = await publish(
      path.join(__dirname, '../../temp/download/axios-0.19.2.tgz'),
      'false',
    );
    expect(data).toMatchInlineSnapshot(`true`);
  });
});

describe('npm view', () => {
  describe.each(cross(['npm', 'taobao'], ['axios@0.19.1', 'webpack@4.41.5']))(
    'registry: %s, package: %s',
    (registry, pkg) => {
      it(pkg, async () => {
        setRegistry(registry);
        const data = await view(pkg);
        const dataWithoutDistTags = omit(data, ['dist-tags']);
        expect(dataWithoutDistTags).toMatchSnapshot();
        expect(data['dist-tags']).toBeDefined();
      });
    },
  );
});

describe('npm dist-tags ls', () => {
  it('@babel/core', async () => {
    const data = await distTag.ls('@babel/core');
    expect(data).toMatchInlineSnapshot(`
      Object {
        "bridge6": "6.0.0-bridge.1",
        "latest": "7.8.3",
      }
    `);
  });
});
