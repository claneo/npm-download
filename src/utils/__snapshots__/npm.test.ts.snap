// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm view registry: npm, package: axios@0.19.1 axios@0.19.1 1`] = `
Object {
  "_hasShrinkwrap": false,
  "dependencies": Object {
    "follow-redirects": "1.5.10",
  },
  "devDependencies": Object {
    "bundlesize": "^0.17.0",
    "coveralls": "^3.0.0",
    "es6-promise": "^4.2.4",
    "grunt": "^1.0.2",
    "grunt-banner": "^0.6.0",
    "grunt-cli": "^1.2.0",
    "grunt-contrib-clean": "^1.1.0",
    "grunt-contrib-watch": "^1.0.0",
    "grunt-eslint": "^20.1.0",
    "grunt-karma": "^2.0.0",
    "grunt-mocha-test": "^0.13.3",
    "grunt-ts": "^6.0.0-beta.19",
    "grunt-webpack": "^1.0.18",
    "istanbul-instrumenter-loader": "^1.0.0",
    "jasmine-core": "^2.4.1",
    "karma": "^1.3.0",
    "karma-chrome-launcher": "^2.2.0",
    "karma-coverage": "^1.1.1",
    "karma-firefox-launcher": "^1.1.0",
    "karma-jasmine": "^1.1.1",
    "karma-jasmine-ajax": "^0.1.13",
    "karma-opera-launcher": "^1.0.0",
    "karma-safari-launcher": "^1.0.0",
    "karma-sauce-launcher": "^1.2.0",
    "karma-sinon": "^1.0.5",
    "karma-sourcemap-loader": "^0.3.7",
    "karma-webpack": "^1.7.0",
    "load-grunt-tasks": "^3.5.2",
    "minimist": "^1.2.0",
    "mocha": "^5.2.0",
    "sinon": "^4.5.0",
    "typescript": "^2.8.1",
    "url-search-params": "^0.10.0",
    "webpack": "^1.13.1",
    "webpack-dev-server": "^1.14.1",
  },
  "directories": Object {},
  "dist": Object {
    "fileCount": 42,
    "integrity": "sha512-Yl+7nfreYKaLRvAvjNPkvfjnQHJM1yLBY3zhqAwcJSwR/6ETkanUgylgtIvkvz0xJ+p/vZuNw8X7Hnb7Whsbpw==",
    "npm-signature": "-----BEGIN PGP SIGNATURE-----
Version: OpenPGP.js v3.0.4
Comment: https://openpgpjs.org

wsFcBAEBCAAQBQJeFL6cCRA9TVsSAnZWagAAj0kP/0NPs4Nmu741N8YePwEr
FEdiHcOB58DTnV08vCfDEbFhrXOjtx2zAaTtuhfXkCvYUOfRCkUXf2E64rpx
pvAhUurGuOun9kO+BCzja1nBoNWxPQfTY+1P7ebEZm403+ZkwvojEt0Sxj5x
X9zGuvNrwScK3rbUFwZ91h3Kw7Sk4vxFFTBy5aupb2mZr/LAeD+IKQzqbDFq
lkwQAwcUjeNer4VqRprm/CX/AxFrljxIwvI/1W2ksPiUM/D3AtjQKFAqCPBD
a1BvKKXOU2fMychwrhahJ6/X2sCVr0n3M8189lOiBEv1QIePk37fHJo3wffV
gZrBoJ4UsCrQBzjshJVdTJXteSpGxQcv16H/IXDX+zW596KQsJ8Amd4pA2iL
kCAIcKhjxZLcTAZhRGYieI9xtwUt1Kj5b+ALmQY1+fbKOICBO4a+GELeOaAe
1RCU/uzgPMhA+GiwK0W8gBcPH7gk8u0bGQiNNUka50H9GRlaqgN0Ex4JhMDH
VyhdhGnvttYVJV2myWLEjCVRiev91f6SBpBjcXlzoHQP0gyzu/Zx8knokVEt
+P7yzj0u4j90A8fvd9oRxkDfGLsc0qXFEZyIPpvmXP1Rs0VLRflX7X4C2qQw
ooSGYN9bBajw33dkOGHzOiyZ07oIEWdNyeEqQk3mc6aWUusfVIiawrqp7Rm3
kYmc
=MDJG
-----END PGP SIGNATURE-----
",
    "shasum": "8a6a04eed23dfe72747e1dd43c604b8f1677b5aa",
    "tarball": "https://registry.npmjs.org/axios/-/axios-0.19.1.tgz",
    "unpackedSize": 348347,
  },
  "name": "axios",
  "version": "0.19.1",
}
`;

exports[`npm view registry: npm, package: webpack@4.41.5 webpack@4.41.5 1`] = `
Object {
  "_hasShrinkwrap": false,
  "bin": Object {
    "webpack": "./bin/webpack.js",
  },
  "dependencies": Object {
    "@webassemblyjs/ast": "1.8.5",
    "@webassemblyjs/helper-module-context": "1.8.5",
    "@webassemblyjs/wasm-edit": "1.8.5",
    "@webassemblyjs/wasm-parser": "1.8.5",
    "acorn": "^6.2.1",
    "ajv": "^6.10.2",
    "ajv-keywords": "^3.4.1",
    "chrome-trace-event": "^1.0.2",
    "enhanced-resolve": "^4.1.0",
    "eslint-scope": "^4.0.3",
    "json-parse-better-errors": "^1.0.2",
    "loader-runner": "^2.4.0",
    "loader-utils": "^1.2.3",
    "memory-fs": "^0.4.1",
    "micromatch": "^3.1.10",
    "mkdirp": "^0.5.1",
    "neo-async": "^2.6.1",
    "node-libs-browser": "^2.2.1",
    "schema-utils": "^1.0.0",
    "tapable": "^1.1.3",
    "terser-webpack-plugin": "^1.4.3",
    "watchpack": "^1.6.0",
    "webpack-sources": "^1.4.1",
  },
  "devDependencies": Object {
    "@babel/core": "^7.7.2",
    "@types/node": "^10.12.21",
    "@types/tapable": "^1.0.1",
    "@types/webpack-sources": "^0.1.4",
    "@yarnpkg/lockfile": "^1.1.0",
    "babel-loader": "^8.0.6",
    "benchmark": "^2.1.1",
    "bundle-loader": "~0.5.0",
    "coffee-loader": "^0.9.0",
    "coffeescript": "^2.3.2",
    "coveralls": "^3.0.2",
    "css-loader": "^2.1.0",
    "es6-promise-polyfill": "^1.1.1",
    "eslint": "^5.8.0",
    "eslint-config-prettier": "^4.0.0",
    "eslint-plugin-jest": "^22.2.2",
    "eslint-plugin-jsdoc": "^15.3.2",
    "eslint-plugin-node": "^8.0.0",
    "eslint-plugin-prettier": "^3.0.0",
    "express": "~4.16.4",
    "file-loader": "^3.0.1",
    "glob": "^7.1.3",
    "husky": "^1.1.3",
    "i18n-webpack-plugin": "^1.0.0",
    "istanbul": "^0.4.5",
    "jest": "^24.9.0",
    "jest-junit": "^8.0.0",
    "json-loader": "^0.5.7",
    "json-schema-to-typescript": "^6.0.1",
    "less": "^3.9.0",
    "less-loader": "^4.0.3",
    "lint-staged": "^8.0.4",
    "lodash": "^4.17.4",
    "prettier": "^1.14.3",
    "pug": "^2.0.4",
    "pug-loader": "^2.4.0",
    "raw-loader": "^1.0.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0",
    "rimraf": "^2.6.2",
    "script-loader": "~0.7.0",
    "simple-git": "^1.65.0",
    "strip-ansi": "^5.2.0",
    "style-loader": "^0.23.1",
    "typescript": "^3.0.0-rc",
    "url-loader": "^1.1.2",
    "val-loader": "^1.0.2",
    "vm-browserify": "~1.1.0",
    "wast-loader": "^1.5.5",
    "webpack-dev-middleware": "^3.5.1",
    "worker-loader": "^2.0.0",
    "xxhashjs": "^0.2.1",
  },
  "directories": Object {},
  "dist": Object {
    "fileCount": 357,
    "integrity": "sha512-wp0Co4vpyumnp3KlkmpM5LWuzvZYayDwM2n17EHFr4qxBBbRokC7DJawPJC7TfSFZ9HZ6GsdH40EBj4UV0nmpw==",
    "npm-signature": "-----BEGIN PGP SIGNATURE-----
Version: OpenPGP.js v3.0.4
Comment: https://openpgpjs.org

wsFcBAEBCAAQBQJeBnNACRA9TVsSAnZWagAAop0P/0xSaHYtN5rZDZKkEwLc
VKIU2NwFjoum54HyesbEYraES5APN/rYXU/v8i68pgg8gvSqB+jWyy6I6N69
IaI9DCz9shfmqHgnddjeL8JuNPrSsk8Bw1Fb0s1NrSjOl6Y1dkDnmcvcnRUF
hL9Qt1fJktHyoEpshYejqx/U+kKLsPKS/tcoUhuOhtxlOvlfM+uI0f+xrn/J
i/BHFFjby2pcb8RzIjJ60GAYX19EDk+vvP01oKWkBCYtrrfH0/IaGgUK3N0E
AASfNwDk3qvjBKU0k9DhJYf8TsLs+XXrq5rNVqQIUS305vCpwHMmaSMvlCWP
gyP9NuFL0OKMEc/KhPhT3l1nRZkW1SIQMy3264Q27FPl1iXpAPDWOf7rMLUk
oDmuhlKMwTy8ztzO1RuvHqKDUQSQLdouD6IdbLw2zq8VhnDaHtmFo8TXF1U8
mZc4PBITU0sEjtq0imcXoydMAMZl4P0AZ8Kp8NtG5jOhTcJm0WrNKkpwM9DA
+RhYIRM6MuZfel2zyhzTkwZ/yq0GN2V5UQgYvR7CC9QTOAq/nQU6rGRfexac
vhQfdV/bWzP6wDajTHsTxVWmt7UY1DpMdsbeLuZfntuhr7Ws51HdoAK9yqZr
tjCIaL8wK5Jbj3tqzjqjqn5VdqG0VdwscN2Tf7JjVFQA1mnL26k076SrqVaS
bIlk
=mnqh
-----END PGP SIGNATURE-----
",
    "shasum": "3210f1886bce5310e62bb97204d18c263341b77c",
    "tarball": "https://registry.npmjs.org/webpack/-/webpack-4.41.5.tgz",
    "unpackedSize": 1463888,
  },
  "engines": Object {
    "node": ">=6.11.5",
  },
  "name": "webpack",
  "version": "4.41.5",
}
`;

exports[`npm view registry: taobao, package: axios@0.19.1 axios@0.19.1 1`] = `
Object {
  "_hasShrinkwrap": false,
  "dependencies": Object {
    "follow-redirects": "1.5.10",
  },
  "devDependencies": Object {
    "bundlesize": "^0.17.0",
    "coveralls": "^3.0.0",
    "es6-promise": "^4.2.4",
    "grunt": "^1.0.2",
    "grunt-banner": "^0.6.0",
    "grunt-cli": "^1.2.0",
    "grunt-contrib-clean": "^1.1.0",
    "grunt-contrib-watch": "^1.0.0",
    "grunt-eslint": "^20.1.0",
    "grunt-karma": "^2.0.0",
    "grunt-mocha-test": "^0.13.3",
    "grunt-ts": "^6.0.0-beta.19",
    "grunt-webpack": "^1.0.18",
    "istanbul-instrumenter-loader": "^1.0.0",
    "jasmine-core": "^2.4.1",
    "karma": "^1.3.0",
    "karma-chrome-launcher": "^2.2.0",
    "karma-coverage": "^1.1.1",
    "karma-firefox-launcher": "^1.1.0",
    "karma-jasmine": "^1.1.1",
    "karma-jasmine-ajax": "^0.1.13",
    "karma-opera-launcher": "^1.0.0",
    "karma-safari-launcher": "^1.0.0",
    "karma-sauce-launcher": "^1.2.0",
    "karma-sinon": "^1.0.5",
    "karma-sourcemap-loader": "^0.3.7",
    "karma-webpack": "^1.7.0",
    "load-grunt-tasks": "^3.5.2",
    "minimist": "^1.2.0",
    "mocha": "^5.2.0",
    "sinon": "^4.5.0",
    "typescript": "^2.8.1",
    "url-search-params": "^0.10.0",
    "webpack": "^1.13.1",
    "webpack-dev-server": "^1.14.1",
  },
  "directories": Object {},
  "dist": Object {
    "integrity": "sha1-imoE7tI9/nJ0fh3UPGBLjxZ3tao=",
    "noattachment": false,
    "shasum": "8a6a04eed23dfe72747e1dd43c604b8f1677b5aa",
    "size": 88708,
    "tarball": "https://registry.npm.taobao.org/axios/download/axios-0.19.1.tgz",
  },
  "name": "axios",
  "publish_time": 1578417819782,
  "version": "0.19.1",
}
`;

exports[`npm view registry: taobao, package: webpack@4.41.5 webpack@4.41.5 1`] = `
Object {
  "_hasShrinkwrap": false,
  "bin": Object {
    "webpack": "./bin/webpack.js",
  },
  "dependencies": Object {
    "@webassemblyjs/ast": "1.8.5",
    "@webassemblyjs/helper-module-context": "1.8.5",
    "@webassemblyjs/wasm-edit": "1.8.5",
    "@webassemblyjs/wasm-parser": "1.8.5",
    "acorn": "^6.2.1",
    "ajv": "^6.10.2",
    "ajv-keywords": "^3.4.1",
    "chrome-trace-event": "^1.0.2",
    "enhanced-resolve": "^4.1.0",
    "eslint-scope": "^4.0.3",
    "json-parse-better-errors": "^1.0.2",
    "loader-runner": "^2.4.0",
    "loader-utils": "^1.2.3",
    "memory-fs": "^0.4.1",
    "micromatch": "^3.1.10",
    "mkdirp": "^0.5.1",
    "neo-async": "^2.6.1",
    "node-libs-browser": "^2.2.1",
    "schema-utils": "^1.0.0",
    "tapable": "^1.1.3",
    "terser-webpack-plugin": "^1.4.3",
    "watchpack": "^1.6.0",
    "webpack-sources": "^1.4.1",
  },
  "devDependencies": Object {
    "@babel/core": "^7.7.2",
    "@types/node": "^10.12.21",
    "@types/tapable": "^1.0.1",
    "@types/webpack-sources": "^0.1.4",
    "@yarnpkg/lockfile": "^1.1.0",
    "babel-loader": "^8.0.6",
    "benchmark": "^2.1.1",
    "bundle-loader": "~0.5.0",
    "coffee-loader": "^0.9.0",
    "coffeescript": "^2.3.2",
    "coveralls": "^3.0.2",
    "css-loader": "^2.1.0",
    "es6-promise-polyfill": "^1.1.1",
    "eslint": "^5.8.0",
    "eslint-config-prettier": "^4.0.0",
    "eslint-plugin-jest": "^22.2.2",
    "eslint-plugin-jsdoc": "^15.3.2",
    "eslint-plugin-node": "^8.0.0",
    "eslint-plugin-prettier": "^3.0.0",
    "express": "~4.16.4",
    "file-loader": "^3.0.1",
    "glob": "^7.1.3",
    "husky": "^1.1.3",
    "i18n-webpack-plugin": "^1.0.0",
    "istanbul": "^0.4.5",
    "jest": "^24.9.0",
    "jest-junit": "^8.0.0",
    "json-loader": "^0.5.7",
    "json-schema-to-typescript": "^6.0.1",
    "less": "^3.9.0",
    "less-loader": "^4.0.3",
    "lint-staged": "^8.0.4",
    "lodash": "^4.17.4",
    "prettier": "^1.14.3",
    "pug": "^2.0.4",
    "pug-loader": "^2.4.0",
    "raw-loader": "^1.0.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0",
    "rimraf": "^2.6.2",
    "script-loader": "~0.7.0",
    "simple-git": "^1.65.0",
    "strip-ansi": "^5.2.0",
    "style-loader": "^0.23.1",
    "typescript": "^3.0.0-rc",
    "url-loader": "^1.1.2",
    "val-loader": "^1.0.2",
    "vm-browserify": "~1.1.0",
    "wast-loader": "^1.5.5",
    "webpack-dev-middleware": "^3.5.1",
    "worker-loader": "^2.0.0",
    "xxhashjs": "^0.2.1",
  },
  "directories": Object {},
  "dist": Object {
    "integrity": "sha1-MhDxiGvOUxDmK7lyBNGMJjNBt3w=",
    "noattachment": false,
    "shasum": "3210f1886bce5310e62bb97204d18c263341b77c",
    "size": 308415,
    "tarball": "https://registry.npm.taobao.org/webpack/download/webpack-4.41.5.tgz",
  },
  "engines": Object {
    "node": ">=6.11.5",
  },
  "name": "webpack",
  "publish_time": 1577481023868,
  "version": "4.41.5",
}
`;

exports[`npm view registry: taobao, package: webpack@4.41.5 webpack@4.41.5 2`] = `
Object {
  "_hasShrinkwrap": false,
  "bin": Object {
    "webpack": "./bin/webpack.js",
  },
  "dependencies": Object {
    "@webassemblyjs/ast": "1.8.5",
    "@webassemblyjs/helper-module-context": "1.8.5",
    "@webassemblyjs/wasm-edit": "1.8.5",
    "@webassemblyjs/wasm-parser": "1.8.5",
    "acorn": "^6.2.1",
    "ajv": "^6.10.2",
    "ajv-keywords": "^3.4.1",
    "chrome-trace-event": "^1.0.2",
    "enhanced-resolve": "^4.1.0",
    "eslint-scope": "^4.0.3",
    "json-parse-better-errors": "^1.0.2",
    "loader-runner": "^2.4.0",
    "loader-utils": "^1.2.3",
    "memory-fs": "^0.4.1",
    "micromatch": "^3.1.10",
    "mkdirp": "^0.5.1",
    "neo-async": "^2.6.1",
    "node-libs-browser": "^2.2.1",
    "schema-utils": "^1.0.0",
    "tapable": "^1.1.3",
    "terser-webpack-plugin": "^1.4.3",
    "watchpack": "^1.6.0",
    "webpack-sources": "^1.4.1",
  },
  "devDependencies": Object {
    "@babel/core": "^7.7.2",
    "@types/node": "^10.12.21",
    "@types/tapable": "^1.0.1",
    "@types/webpack-sources": "^0.1.4",
    "@yarnpkg/lockfile": "^1.1.0",
    "babel-loader": "^8.0.6",
    "benchmark": "^2.1.1",
    "bundle-loader": "~0.5.0",
    "coffee-loader": "^0.9.0",
    "coffeescript": "^2.3.2",
    "coveralls": "^3.0.2",
    "css-loader": "^2.1.0",
    "es6-promise-polyfill": "^1.1.1",
    "eslint": "^5.8.0",
    "eslint-config-prettier": "^4.0.0",
    "eslint-plugin-jest": "^22.2.2",
    "eslint-plugin-jsdoc": "^15.3.2",
    "eslint-plugin-node": "^8.0.0",
    "eslint-plugin-prettier": "^3.0.0",
    "express": "~4.16.4",
    "file-loader": "^3.0.1",
    "glob": "^7.1.3",
    "husky": "^1.1.3",
    "i18n-webpack-plugin": "^1.0.0",
    "istanbul": "^0.4.5",
    "jest": "^24.9.0",
    "jest-junit": "^8.0.0",
    "json-loader": "^0.5.7",
    "json-schema-to-typescript": "^6.0.1",
    "less": "^3.9.0",
    "less-loader": "^4.0.3",
    "lint-staged": "^8.0.4",
    "lodash": "^4.17.4",
    "prettier": "^1.14.3",
    "pug": "^2.0.4",
    "pug-loader": "^2.4.0",
    "raw-loader": "^1.0.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0",
    "rimraf": "^2.6.2",
    "script-loader": "~0.7.0",
    "simple-git": "^1.65.0",
    "strip-ansi": "^5.2.0",
    "style-loader": "^0.23.1",
    "typescript": "^3.0.0-rc",
    "url-loader": "^1.1.2",
    "val-loader": "^1.0.2",
    "vm-browserify": "~1.1.0",
    "wast-loader": "^1.5.5",
    "webpack-dev-middleware": "^3.5.1",
    "worker-loader": "^2.0.0",
    "xxhashjs": "^0.2.1",
  },
  "directories": Object {},
  "dist": Object {
    "integrity": "sha1-MhDxiGvOUxDmK7lyBNGMJjNBt3w=",
    "noattachment": false,
    "shasum": "3210f1886bce5310e62bb97204d18c263341b77c",
    "size": 308415,
    "tarball": "https://registry.npm.taobao.org/webpack/download/webpack-4.41.5.tgz",
  },
  "engines": Object {
    "node": ">=6.11.5",
  },
  "name": "webpack",
  "publish_time": 1577481023868,
  "version": "4.41.5",
}
`;
